### PR TITLE
Add an Int63 module

### PR DESCRIPTION
**Pull request by craigfe**
This adds a new top-level `Int63` module to this library, which is implemented by `int` on 64-bit architectures and `int64` on 32-bit architectures. This new module has exactly the same interface as `Optint` (which now contains 64-bit conversion functions too). The boxed implementation is wrapped in order to ensure that it keeps the same 63-bit semantics as regular integers, at the cost of some performance.

The two `Int63` implementations are tested to be equivalent using [`monolith`](https://gitlab.inria.fr/fpottier/monolith/).

I've also re-written the `README.md` to add a bit more context and to reflect the stability of the library.

## TODO

- add efficient `encode` / `decode` functions for this type, which will be needed to support the [use-case in `index`](https://github.com/CraigFe/index/blob/fanout-63-bit-integers/src/fan.ml#L71-L89). These should be cross-compatible with the codecs for `Int64`.
- add the new fuzz tests to the CI in some way.

### CODE REVIEW

File: fuzz/fuzz_int63.ml [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+    declare "compare" (t ^> t ^> int) R.compare C.compare;
+    declare "equal" (t ^> t ^> bool) R.equal C.equal;
+
+    declare "of_int" (int ^> t) R.of_int C.of_int;
+    declare "to_int" (t ^> int) R.to_int C.to_int;
+    declare "of_int32" (int32 ^> t) R.of_int32 C.of_int32;
+    declare "to_int32" (t ^> int32) R.to_int32 C.to_int32;
+    declare "to_float" (t ^> float) R.to_float C.to_float;
+    declare "to_string" (t ^> string) R.to_string C.to_string;
+
```
Comment (Ngoguey42) [inline fuzz/fuzz_int63.ml:89]: You missed `of_string`, `of_string_opt` and `pp`! :)

**Review by Ngoguey42 — COMMENTED** [review]
Super great

File: fuzz/fuzz_int63.ml (hunk 1 above)
Comment (craigfe) [reply fuzz/fuzz_int63.ml:89]: Those were all missed intentionally because they're unpleasant to fuzz, but I can workaround that :-) If nothing else, `to_string >> of_string` should be the identity.

Comment (Ngoguey42) [reply fuzz/fuzz_int63.ml:89]: Don't bother! :D

**Comment by Ngoguey42** [comment]
Where `let mul x y = if x = minus_one then x else Int64.mul x (Conv.unwrap y)` would crash immediately, `let mul x y = if x = 123L && y = 321L then x else Int64.mul x (Conv.unwrap y)` ran all night without errors. 
```
  8G tests run so far (237K/s overall, 241K/s now) (fuel = 5).
```

I was expecting afl to catch such an edge case.

**Comment by craigfe** [comment]
As I understand it, this is a weakness of both Crowbar and Monolith: they don't know the correspondence between their own generators and internal branches. So with a generator in the range `[int_min, int_max]` there is no special affinity towards picking values that explore different branches in the code. Values that are easily reachable from the interface itself (`int_min`, `int_max`, powers of two etc.) should be well-covered however – many of the counter-examples I got while developing used these values.

**Comment by dinosaure** [comment]
I added OCaml-CI on this repository and I will fix outcome issues. The PR seems good enough. I will check by hands reverse dependencies and cut a release then. Thanks :+1:.

**Comment by craigfe** [comment]
I've added some proposed encode / decode functions now. Let me know if you'd like any changes / would like me to help with getting the CI up and running :-)

**Comment by dinosaure** [comment]
As far as I can tell, a reverse dependencies on `decompress` and `ocaml-git` does not show any bugs on `x86` and `x64`. So, I believe that this PR is fine :+1:.
